### PR TITLE
PAT-1594: ReviewStep ImageCapture full screen image

### PR DESCRIPTION
**Pull request by AbdallahMedable**
### Objective
- Added implementation for allowing captured images to display in full screen.

### How To Test
**Credentials:**
- Environment: Int-Dev
- Org: abdallahandela
- Study: Banadol

**Steps:**
  1. Open the app and use the credentials above
  2. Join Study
  3. Take task `ReviewStep ImageCapture`
  4. Reach to ReviewStep

**Expected:**
There should be an expand button for each image that opens the image in a new screen.
The new screen should have the image, a close button that returns to ReviewStep, and an Edit button that takes us back to editing the corresponding step.

See the design:
- https://app.zeplin.io/project/5d8e16c2db3c0d61aac43a5f/screen/5da0bfe52eea29085f322710
- https://app.zeplin.io/project/5d8e16c2db3c0d61aac43a5f/screen/5d9e23880975412f90b00111

### Branches
- PAT: development
- Axon: task/PAT-1594_ReviewStep_ImageCapture_FullScreen
- RS: task/PAT-1594_ReviewStep_ImageCapture_FullScreen
- Cortex: development

### Links
- https://jira.devops.medable.com/browse/PAT-1594

Closes PAT-1594

### CODE REVIEW

**Comment by AbdallahMedable** [comment]
1. The tiny pictures are yet to be investigated, it's either something device-specific, or there's something wrong in our current implementation for RSImageCaptureLayout
2. Applied all UI enhancements
3. This ticket isn't editing the ImageCapture in any way, it's only adding a full screen view for captured images.

File: backbone/src/main/java/org/researchstack/backbone/ui/step/fragments/ReviewStepFragment.kt [hunk 1]
```diff
@@ -1,15 +1,24 @@
...
 internal class ReviewStepFragment : BaseStepFragment(R.layout.rsb_fragment_review_step_layout),
-        ReviewStepRowClickHandler {
+                                    ReviewStepRowClickHandler {
+    companion object {
+        private const val REQUEST_FULL_SCREEN_VIEW = 101
+        const val RESULT_FULL_SCREEN_EDIT = 201
+        const val EXTRA_FULL_SCREEN_STEP = "full_screen_step"
+    }
+
+    var stepToEdit: Step? = null
```
Comment (Gryzor) [inline backbone/src/main/java/org/researchstack/backbone/ui/step/fragments/ReviewStepFragment.kt]: can't this be private?

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    init {
+        loadImage()
+    }
+
+    private fun loadImage() {
+        imageUrl?.let {
+            var bitmap: Bitmap? = null
+            try {
+                bitmap = getBitmapFromFile(it)
```
Comment (Gryzor) [inline backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageViewModel.kt]: Can't this throw an OOM exception if the bitmap to decode is big enough?

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        android:background="?android:attr/selectableItemBackground"
+        android:contentDescription="@string/rsb_edit_step"
+        android:drawableStart="@drawable/rsb_ic_edit"
+        android:drawablePadding="10dp"
+        android:drawableTint="@android:color/white"
+        android:text="@string/rsb_edit_step"
+        android:textAllCaps="true"
+        android:textColor="@android:color/white"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/fullScreenImage"
```
Comment (Gryzor) [inline backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: What happens if the image (whose bottom is pinned to the bottom of the parent) effectively occupies the full screen? This textView is going to be off-screen and not seen.

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        app:layout_constraintTop_toTopOf="parent"
+        tools:ignore="HardcodedText" />
+
+    <TextView
+        android:id="@+id/fullScreenImageEditStep"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="20dp"
+        android:layout_marginEnd="16dp"
+        android:background="?android:attr/selectableItemBackground"
```
Comment (Gryzor) [inline backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: I think this should be foreground, not background. You want the effect above the content.

File: backbone/src/main/res/layout/rsb_image_capture_step_layout.xml [hunk 5]
```diff
@@ -47,5 +47,14 @@
...
         app:srcCompat="@drawable/rsb_ic_circle_8dp" />
 
+    <ImageButton
+        android:id="@+id/imageCaptureImageStepQuestionAnswerFullscreen"
+        style="@style/rsb_image_capture_fullscreen_style"
+        app:layout_constraintEnd_toEndOf="@+id/imageCaptureImageStepQuestionAnswer"
+        android:contentDescription="Full Screen"
+        android:background="@null"
+        app:layout_constraintTop_toTopOf="@+id/imageCaptureImageStepQuestionAnswer"
+        app:srcCompat="@drawable/rsb_ic_review_step_image_fullscreen" />
```
Comment (Gryzor) [inline backbone/src/main/res/layout/rsb_image_capture_step_layout.xml]: You're using a PNG, any reason?

I would remove all this, and use the included in android (you have to import it from the Vector Graphics) "Crop Free" svg. It looks the same, it's a vector graphics, etc. 

Remove the imported PNGs.

File: backbone/src/main/res/values/styles.xml [hunk 6]
```diff
@@ -138,6 +138,14 @@
         <item name="android:layout_width">100dp</item>
         <item name="android:layout_height">100dp</item>
         <item name="android:layout_marginStart">16dp</item>
+        <item name="android:scaleType">centerCrop</item>
```
Comment (Gryzor) [inline backbone/src/main/res/values/styles.xml]: What does this fix exactly?

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+
+    private fun initCallbacks() {
+        viewModel.displayImageEvent.observe(this, Observer {
+            fullScreenImage.setImageBitmap(it)
+        })
+        fullScreenImageClose.setOnClickListener(this)
+        fullScreenImageEditStep.setOnClickListener(this)
+    }
+
+    override fun onClick(v: View?) {
```
Comment (Gryzor) [inline backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: if you don't mind and because you're going to be making changes, can you rename this "v" to `view`.

**Comment by jaimesDSG** [comment]
Maybe the image on the review is by default a little bit small ? 

image

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt (hunk 7 above)
Comment (jaimesDSG) [reply backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: Personnally i prefer to avoid to do the onClick like that. It's to avoid to always to the when stuff.

I prefer for example to do something like that : 

fullScreenImageClose.setOnClickListener {
    setResult(Activity.RESULT_CANCELED)
    finish()
}
fullScreenImageEditStep.setOnClickListener {
    val step: Step? = intent.extras?.getSerializable(EXTRA_STEP) as Step?
    val data = Intent()
    data.putExtra(ReviewStepFragment.EXTRA_FULL_SCREEN_STEP, step)
    setResult(ReviewStepFragment.RESULT_FULL_SCREEN_EDIT, data)
    finish()
}

**Comment by Gryzor** [comment]
> Maybe the image on the review is by default a little bit small ?

The Images on the Emulator are (as far as we have observed so far with @norberto-app ) **always small**.

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt (hunk 7 above)
Comment (Gryzor) [reply backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: I think both methods are fine, when you have to set up a bunch of click listeners, using "this" is preferred because in your example, you're creating two anonymous instances of the ClickListener interface, whereas if you set a common one, you only need one reference to the existing class _polymorphed_ as a ClickListener. 

Additionally, when they are related like in this case, it's a lot easier to find the code next to each other; I agree that it really depends. In this case, I like how he did it rather than individual ones, but I have no argument against either besides the irrelevant memory footprint and legibility in some cases. It usually also depends what you do inside the click listener. 

Personally, if I had to set up a bunch of click listeners and I wanted them inline/individually, I'd rather use method reference:

`view.setOnClickListener(this::doSomething)`
`view2.setOnClickListener(this::doSomethingElse)`

And have 
```
    private fun doSomething(view: View) {

    }

    private fun doSomethingElse(view: View) {

    }
```

Or alternatively: 

```
view.setOnClickListener(this::handleSomething)
view2.setOnClickListener(this::handleSomething)

    private fun handleSomething(view: View) {
      when (view..) 
            R.id.thisView -> doThis()
            R.id.thatView -> doThat()
    }
```

In short: it really depends :) What are your thoughts?

**Comment by jaimesDSG** [comment]
> > Maybe the image on the review is by default a little bit small ?
> 
> The Images on the Emulator are (as far as we have observed so far with @norberto-app ) **always small**.

I tested it on a real device and it's the same size for all images on ReviewStep (less than 1/3 of the screen width)

File: backbone/src/main/res/values-ja-rJP/strings.xml [hunk 8]
```diff
@@ -160,4 +160,5 @@
     <string name="rsb_edit_step_alert_step_skip_content">[ja-rJP] Your original response will be discarded..</string>
     <string name="rsb_edit_step_alert_step_skip_discard">[ja-rJP] NO, DON\'T SKIP</string>
     <string name="rsb_edit_step_alert_step_skip_positive">[ja-rJP] Yes, skip</string>
+    <string name="rsb_edit_step">[ja-rJP] Edit</string>
```
Comment (jaimesDSG) [inline backbone/src/main/res/values-ja-rJP/strings.xml]: Maybe it's the same issue for all these strings but why not [ja-JP] ?

File: backbone/src/main/res/values/styles.xml (hunk 6 above)
Comment (jaimesDSG) [reply backbone/src/main/res/values/styles.xml]: Normally (and you can tell me if i am wrong) this will avoid to resize the capture. You will only get the centre and crop everything else. Certainly to avoid distorted image ?

Comment (Gryzor) [reply backbone/src/main/res/values/styles.xml]: @jaimesDSG I won't correct you because you're not wrong :) For the record, I know what centerCrop does, and what all the other scaleTypes do. My question is: what is this fixing *here*. This line is added with no mention in the PR description, so I'm asking: why is this needed, what is this fixing, and why CROPPING (therefore LOSING) part of a picture?

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml (hunk 3 above)
Comment (AbdallahMedable) [reply backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: If I set the image to fill screen, the close and edit buttons may disappear. and if I bind it to the close and edit buttons, the edit button will always be at the bottom of the screen, not right below the image.
What do you suggest we do here?

File: backbone/src/main/res/values/styles.xml (hunk 6 above)
Comment (AbdallahMedable) [reply backbone/src/main/res/values/styles.xml]: I'm cropping this because I need to add the tiny square icon in the top right corner, If I don't crop, I won't be able to know the exact view size and the square icon will not be aligned correctly. if you have a better way to do this please let me know

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt (hunk 7 above)
Comment (AbdallahMedable) [reply backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: I usually use `setOnClickListener(this)` when there are several views to click. It's exactly what Martin said about having one instance instead of having several.

File: backbone/src/main/res/values-ja-rJP/strings.xml (hunk 8 above)
Comment (AbdallahMedable) [reply backbone/src/main/res/values-ja-rJP/strings.xml]: I'm just using the file name postfix `values-ja-rJP`

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml (hunk 3 above)
Comment (jaimesDSG) [reply backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: It's not possible to have button that "may disappear" i think.

File: backbone/src/main/res/values-ja-rJP/strings.xml (hunk 8 above)
Comment (jaimesDSG) [reply backbone/src/main/res/values-ja-rJP/strings.xml]: I understand but for all the other when it's [fr-rFR] we add the prefix [fr-FR] not [fr-rFR] ^^ But i think it's not really important

**Comment by AbdallahMedable** [comment]
Put as WIP until UI team replies on some questions I posted on zeplin

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml (hunk 3 above)
Comment (AbdallahMedable) [reply backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: I just fixed that in my latest commit, but I'm still waiting for UI team to get back to me on how to resolve certain cases for this image (i.e. if it gets bigger, or smaller for any reason)

Comment (AbdallahMedable) [reply backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: According to Felix from UI, The edit button should follow the height of the image. I did this, and made sure if the Image is going to fill the screen, the edit button would still be visible

File: backbone/src/main/res/values/styles.xml (hunk 6 above)
Comment (AbdallahMedable) [reply backbone/src/main/res/values/styles.xml]: So I [asked](https://app.zeplin.io/project/5d8e16c2db3c0d61aac43a5f/screen/5da0bfe52eea29085f322710) Felix from UI on Zeplin on the best way to show the maximize button, and we agreed I should crop the image and fill the view size

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="10dp"
+        android:layout_marginEnd="10dp"
+        android:contentDescription="Image"
+        android:scaleType="centerInside"
+        app:layout_constraintBottom_toTopOf="@id/fullScreenImageEditStep"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
```
Comment (Gryzor) [inline backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: Shouldn't this be _below_ `fullScreenImageClose` ? If the image is taller than the screen, this will pin to the top, sitting on top of the close button or am I wrong?

File: backbone/src/main/res/layout/rsb_image_capture_step_layout.xml [hunk 10]
```diff
@@ -47,5 +47,14 @@
         app:layout_constraintTop_toBottomOf="@+id/imageCaptureStepQuestionItself"
         app:srcCompat="@drawable/rsb_ic_circle_8dp" />
 
+    <ImageButton
+        android:id="@+id/imageCaptureImageStepQuestionAnswerFullscreen"
+        style="@style/rsb_image_capture_fullscreen_style"
+        app:layout_constraintEnd_toEndOf="@+id/imageCaptureImageStepQuestionAnswer"
+        android:contentDescription="Full Screen"
```
Comment (Gryzor) [inline backbone/src/main/res/layout/rsb_image_capture_step_layout.xml]: Hardcoded string here. This should go to the multi-language spreadsheet and become a string.

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    private fun initCallbacks() {
+        viewModel.displayImageEvent.observe(this, Observer {
+            fullScreenImage.setImageBitmap(it)
+        })
+        fullScreenImageClose.setOnClickListener(this)
+        fullScreenImageEditStep.setOnClickListener(this)
+
+        fullScreenImageContainer.viewTreeObserver.addOnGlobalLayoutListener(object :
```
Comment (Gryzor) [inline backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: Chris Bane (from Google) once said: 99% of the times, using viewTreeObserver is "massively overkill" ([source](https://twitter.com/chrisbanes/status/986171606077550592)). Can you try to see if there's a simpler/better way to do this? Perhaps move the code onStart? If not, this is fine, although *if the project has KTX-Extensions* (can't remember now), perhaps you can try to see if the android-ktx extension `doOnLayout()` or similar is enough? That'd onlyl be applied to your view (as opposed to the entire layout).

File: backbone/src/main/res/layout/rsb_image_capture_step_layout.xml (hunk 10 above)
Comment (AbdallahMedable) [reply backbone/src/main/res/layout/rsb_image_capture_step_layout.xml]: I was thinking when this MR is merged, I'll add this string to the Accessibility MR that's been blocked for a while now

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt (hunk 11 above)
Comment (AbdallahMedable) [reply backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: We aren't using ktx in this project, so I'm gonna use `addOnLayoutChangeListener` instead.

File: backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt (hunk 7 above)
Comment (Gryzor) [reply backbone/src/main/java/org/researchstack/backbone/ui/task/ReviewStepFullScreenImageActivity.kt]: And can you rename it to `view` instead of `v`? :) Thanks!

(that was the initial comment)

File: backbone/src/main/res/layout/rsb_activity_full_screen_image.xml [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+    <ImageButton
+        android:id="@+id/fullScreenImageClose"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="16dp"
+        android:layout_marginEnd="16dp"
+        android:background="@null"
+        android:contentDescription="Close"
+        android:src="@drawable/rsb_ic_close"
```
Comment (Gryzor) [inline backbone/src/main/res/layout/rsb_activity_full_screen_image.xml]: Can you replace this `.png` with the equivalent vector asset please? If it's just the "white" close, Android has one in the VectorAsset library for material design from Android Studio.

File: backbone/src/main/res/layout/rsb_image_capture_step_layout.xml (hunk 10 above)
Comment (Gryzor) [reply backbone/src/main/res/layout/rsb_image_capture_step_layout.xml]: Ok.

Comment (Gryzor) [reply backbone/src/main/res/layout/rsb_image_capture_step_layout.xml]: If you forget, you _pay the beers_
